Sources/Penny/SOChecker.swift

```diff
@@ -0,0 +1,147 @@
+import Logging
+import DiscordBM
+import Markdown
+import Foundation
+
+actor SOChecker {
+
+    struct Storage: Sendable, Codable {
+        var lastCheckDate: Date? = nil
+    }
+
+    var storage = Storage()
+
+    let soService: any SOService
+    let discordService: DiscordService
+    let logger = Logger(label: "SOChecker")
+
+    init(soService: any SOService, discordService: DiscordService) {
+        self.soService = soService
+        self.discordService = discordService
+    }
+
+    nonisolated func run() {
+        Task { [self] in
+            do {
+                try await self.check()
+                try await Task.sleep(for: .seconds(60 * 5)) /// 5 mins
+            } catch {
+                logger.report("Couldn't check SO questions", error: error)
+                try await Task.sleep(for: .seconds(60 * 5))
+            }
+            self.run()
+        }
+    }
+
+    func check() async throws {
+        let after = storage.lastCheckDate ?? Date().addingTimeInterval(-60 * 60)
+        let questions = try await soService.listQuestions(after: after)
+        storage.lastCheckDate = Date()
+
+        for question in questions {
+            await discordService.sendMessage(
+                channelId: Constants.Channels.stackOverflow.id,
+                payload: .init(embeds: [.init(
+                    title: question.title.htmlDecoded().unicodesPrefix(256),
+                    url: question.link,
+                    timestamp: Date(timeIntervalSince1970: Double(question.creationDate)),
+                    color: .mint,
+                    footer: .init(
+                        text: "By \(question.owner.displayName)",
+                        icon_url: question.owner.profileImage.map { .exact($0) }
+                    )
+                )])
+            )
+        }
+    }
+
+    func consumeCachesStorageData(_ storage: Storage) {
+        self.storage = storage
+    }
+
+    func getCachedDataForCachesStorage() -> Storage {
+        return self.storage
+    }
+}
+
+// MARK: +String
+private extension String {
+    func htmlDecoded() -> String {
+        Document(parsing: self).format()
+    }
+}
+
+// MARK: - SOQuestions
+struct SOQuestions: Codable {
+
+    struct Item: Codable {
+
+        struct Owner: Codable {
+            let accountID: Int?
+            let reputation: Int?
+            let userID: Int?
+            let userType: String
+            let acceptRate: Int?
+            let profileImage: String?
+            let displayName: String
+            let link: String?
+
+            enum CodingKeys: String, CodingKey {
+                case accountID = "account_id"
+                case reputation
+                case userID = "user_id"
+                case userType = "user_type"
+                case acceptRate = "accept_rate"
+                case profileImage = "profile_image"
+                case displayName = "display_name"
+                case link
+            }
```

Comment:
Suggestion - set `decoder.keyDecodingStrategy = .convertFromSnakeCase` on the `JSONDecoder` instead of manually specifying this.

Reply:
honestly i like this more for some reason 🤔 maybe i shouldn't